File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, arr: List[int], k: int) -> List[int]:
        n=len(arr)
        list=[]
        dq=[]
        for i in range(n):
            if dq and dq[0]<=i-k:
                dq.pop(0)
            while dq and arr[dq[-1]]<=arr[i]:
                dq.pop()
            dq.append(i)
            if(i>=k-1):
                list.append(arr[dq[0]])
        return list

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] arr, int k) {
        int n=arr.length;
        Deque<Integer> dq=new ArrayDeque<>();
        int[] list=new int[n-k+1];
        int ind=0;
        for (int i=0;i<n;i++){
            if(!dq.isEmpty() && dq.peekFirst()<=i-k) dq.pollFirst();
            while(!dq.isEmpty() && arr[dq.peekLast()]<=arr[i]) dq.pollLast();
            dq.addLast(i);
            if(i>=k-1) list[ind++]=arr[dq.peekFirst()];
        }
        return list;
    }
}